VSD - pointers to heap allocations

jezhiggins opened the pull request:
Extends the variable sensitivity domain to be able to track heap allocated memory.

Within the goto-programs the heap allocation instruction is `ALLOCATE(malloc_size, 0 != 0)`. Within the vsd, this is an expression of type `ID_side_effect` with a statement of `ID_allocate`. When we eval this expression, we create a pointer object with a 'heap-object' placeholder expression. There's a little song and dance to propagate through the pointer type casts returns from malloc are subject to, but we can return 'heap-object' pointer back out.

When we assign pointer to a heap allocation to its symbol, we now know that actual (C) type of the object. We replace the 'heap object' placeholder with an array abstract object. The pointer abstract object can now act exactly as normal.

jezhiggins committed:
Capture heap allocations

Create a special 'heap allocation' placeholder expression. Extend
write_stack to be able to store that expression.

jezhiggins committed:
Allow void* pointers cast to another pointer type

That's the expr.type() rather than the representation class.

jezhiggins committed:
Create heap allocation abstract object on pointer typecast

We can't ensure the modified pointer object will be properly assigned
back to its symbol if we wait until first write.

If we defer then modified pointer may, or may not, be assigned depending
on how the write is expressed

 *p = 1;   // good :)
 p[0] = 1; // bad :(

jezhiggins committed:
heap object index expression needs to be signed_size_type

jezhiggins committed:
Index expression should used signed_size_type rather than size_type

I based my code in the previous commit on this code, but then found it
caused a type error. This leads me to believe that while this current
code path may not be called in normal usage, it is in fact a latent bug.
I believe this change is correct.

martin-cs commented on src/analyses/variable-sensitivity/abstract_pointer_object.cpp:
```diff
@@ -87,6 +100,16 @@ abstract_object_pointert abstract_pointer_objectt::write_dereference(
+  const abstract_environmentt &environment,
+  const namespacet &ns) const
+{
+  INVARIANT(is_void_pointer(type()), "Only allow pointer casting from void*");
```
I feel kind of conflicted about this.  In some ways it should not be an `INVARIANT` because it can be triggered by a user but also, at the moment we really aren't handling this case.  Could we return a `TOP` pointer of the correct type?  Reads from that should give `TOP` and writes to it could call `havoc()`?

jezhiggins replied:
I am similarly conflicted. Right now, the call to typecast() is itself guarded so that it will never be called if this invariant will fire

```
  if(expr.id() == ID_typecast)
  {
    const typecast_exprt &tce = to_typecast_expr(expr);
    if(tce.op().id() == ID_symbol && is_void_pointer(tce.op().type()))
``` 

Consequently existing typecasts, of which there are a handful, will evaluate as they do at the moment. I suppose this invariant is more of warning for future care than a guard against wacky corner cases.

martin-cs commented on src/analyses/variable-sensitivity/constant_pointer_abstract_object.cpp:
```diff
@@ -201,14 +209,41 @@ abstract_object_pointert constant_pointer_abstract_objectt::write_dereference(
+  const abstract_environmentt &environment,
+  const namespacet &ns) const
+{
+  INVARIANT(is_void_pointer(type()), "Only allow pointer casting from void*");
```
As above.

martin-cs commented on src/analyses/variable-sensitivity/constant_pointer_abstract_object.cpp:
```diff
@@ -201,14 +209,41 @@ abstract_object_pointert constant_pointer_abstract_objectt::write_dereference(
+    auto heap_array_type = array_typet(new_type.subtype(), nil_exprt());
+    auto array_object =
+      environment.abstract_object_factory(heap_array_type, ns, true, false);
+    auto heap_symbol = symbol_exprt("heap-object", heap_array_type);
```
Does this potentially clash heap objects of the same type?

jezhiggins replied:
Yes, and I'm working on that now.

jezhiggins replied:
Each heap allocation gets its own symbol. Right now they just go 0, 1, 2, etc, but we could plug in something more sophisticated.

martin-cs commented on src/analyses/variable-sensitivity/value_set_pointer_abstract_object.cpp:
```diff
@@ -133,6 +142,15 @@ abstract_object_pointert value_set_pointer_abstract_objectt::write_dereference(
+  const abstract_environmentt &environment,
+  const namespacet &ns) const
+{
+  INVARIANT(is_void_pointer(type()), "Only allow pointer casting from void*");
```
I see why but can this be left to the implementations of the objects in the set?

jezhiggins replied:
This code is placeholder at the moment. More to follow.

jezhiggins replied:
When we cast a value set of pointers, we need to also cast pointers it contains. We do this now.

jezhiggins committed:
Create symbol name on heap allocation

jezhiggins committed:
Move heap object creation off into object factory.

This feels like the right place for this, and also gives future options
for configuration/customisation.

jezhiggins committed:
Create new symbol for each heap-allocation

jezhiggins committed:
remove redundant apply_comb template

jezhiggins committed:
Typecasting value-set of pointers needs to propagate down

jezhiggins committed:
heap allocation and value-set tests

jezhiggins committed:
vsp = nondet ? malloc(a) : malloc(b) works correctly
